app/Image/SizeVariantDefaultFactory.php

```diff
@@ -0,0 +1,361 @@
+	 * @param float $framePosition the position in seconds
+	 *
+	 * @throws \RuntimeException thrown, if FFmpeg failed to extract a frame
+	 */
+	protected function extractFrame(Video $video, float $framePosition): void
+	{
+		$errMsg = 'Failed to extract snapshot from video ' . $this->referenceFullPath . ' at position ' . $framePosition;
+		try {
+			$frame = $video->frame(TimeCode::fromSeconds($framePosition));
+			$frame->save($this->referenceFullPath);
+		} catch (\RuntimeException $e) {
+			Logs::error(__METHOD__, __LINE__, $errMsg);
+			throw new \RuntimeException($errMsg, 0, $e);
+		}
+		if (!file_exists($this->referenceFullPath) || filesize($this->referenceFullPath) == 0) {
+			throw new \RuntimeException($errMsg);
+		}
+		if (Configs::get_value('lossless_optimization')) {
+			ImageOptimizer::optimize($this->referenceFullPath);
+		}
+	}
+
+	/**
+	 * Creates a temporary path to store the extracted reference image.
+	 *
+	 * This method modifies `referenceFullPath` and also set `needsCleanup`
+	 * to true such that the file which is stored at `referenceFullPath` will
+	 * be removed by {@link SizeVariantFactory::cleanup()}.
+	 */
+	protected function createTmpPathForReference(): void
+	{
+		$this->referenceFullPath = Helpers::createTemporaryFile('.jpeg');
+		$this->needsCleanup = true;
+		Logs::notice(__METHOD__, __LINE__, 'Saving JPG of raw/video file to ' . $this->referenceFullPath);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function createSizeVariants(): Collection
+	{
+		$allVariants = [
+			SizeVariant::THUMB,
+			SizeVariant::THUMB2X,
+			SizeVariant::SMALL,
+			SizeVariant::SMALL2X,
+			SizeVariant::MEDIUM,
+			SizeVariant::MEDIUM2X,
+		];
+		$collection = new Collection();
+
+		foreach ($allVariants as $variant) {
+			$sv = $this->createSizeVariantCond($variant);
+			if ($sv) {
+				$collection->add($sv);
+			}
+		}
+
+		return $collection;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function createSizeVariant(int $sizeVariant): SizeVariant
+	{
+		if ($sizeVariant === SizeVariant::ORIGINAL) {
+			throw new \InvalidArgumentException('createSizeVariant() must not be used to create original size, use createOriginal() instead');
+		}
+		if (empty($this->referenceFullPath)) {
+			$this->extractReferenceImage();
+		}
+		list($maxWidth, $maxHeight) = $this->getMaxDimensions($sizeVariant);
+
+		return $this->createSizeVariantInternal(
+			$sizeVariant, $maxWidth, $maxHeight
+		);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function createSizeVariantCond(int $sizeVariant): ?SizeVariant
+	{
+		if ($sizeVariant === SizeVariant::ORIGINAL) {
+			throw new \InvalidArgumentException('createSizeVariantCond() must not be used to create original size, use createOriginal() instead');
+		}
+		if (!$this->isEnabledByConfiguration($sizeVariant)) {
+			return null;
+		}
+		if (empty($this->referenceFullPath)) {
+			$this->extractReferenceImage();
+		}
+
+		list($maxWidth, $maxHeight) = $this->getMaxDimensions($sizeVariant);
+
```

Comment:
I forget if we ever documented it but the current code does have this feature that if _both_ `$maxWidth` and `$maxHeight` are `0`, the given variant is skipped. I think we should keep that functionality although I'm not sure if it should be an additional check here or whether you prefer it in `isEnabledByConfiguration()`?

Reply:
Obviously, I missed that part of the code. I added an additional test to `isEnabledByConfiguration()`.